components/profile-controller/pkg/controller/profile/profile_controller.go

```diff
@@ -201,6 +222,145 @@ func (r *ReconcileProfile) Reconcile(request reconcile.Request) (reconcile.Resul
+func (r *ReconcileProfile) updateIstioRbac(profileIns *kubeflowv1alpha1.Profile) error {
+	istioServiceRole := &istiorbac.ServiceRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ServiceRoleIstio,
+			Namespace: profileIns.Name,
+		},
+		Spec: istiorbac.ServiceRoleSpec{
+			Rules: []*istiorbac.AccessRule{
+				{
+					Services: []string{"*"},
+				},
+			},
+		},
+	}
+	if err := controllerutil.SetControllerReference(profileIns, istioServiceRole, r.scheme); err != nil {
+		return err
+	}
+	foundSr := &istiorbac.ServiceRole{}
+	err := r.Get(context.TODO(), types.NamespacedName{Name: istioServiceRole.Name,
+		Namespace: istioServiceRole.Namespace}, foundSr)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("Creating Istio ServiceRole", "namespace", istioServiceRole.Namespace,
+				"name", istioServiceRole.Name)
+			err = r.Create(context.TODO(), istioServiceRole)
+			if err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	} else {
+		if !reflect.DeepEqual(istioServiceRole.Spec, foundSr.Spec) {
+			foundSr.Spec = istioServiceRole.Spec
+			log.Info("Updating Istio ServiceRole", "namespace", istioServiceRole.Namespace,
+				"name", istioServiceRole.Name)
+			err = r.Update(context.TODO(), foundSr)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	istioServiceRoleBinding := &istiorbac.ServiceRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ServiceRoleBindingIstio,
+			Namespace: profileIns.Name,
+		},
+		Spec: istiorbac.ServiceRoleBindingSpec{
+			Subjects: []*istiorbac.Subject{
+				{
+					Properties: map[string]string{"request.headers[x-goog-authenticated-user-email]": "accounts.google.com:" + profileIns.Spec.Owner.Name},
+				},
+			},
+			RoleRef: &istiorbac.RoleRef{
+				Kind: "ServiceRole",
+				Name: ServiceRoleIstio,
+			},
+		},
+	}
+	if err := controllerutil.SetControllerReference(profileIns, istioServiceRoleBinding, r.scheme); err != nil {
+		return err
+	}
+	foundSrb := &istiorbac.ServiceRoleBinding{}
+	err = r.Get(context.TODO(), types.NamespacedName{Name: istioServiceRoleBinding.Name,
+		Namespace: istioServiceRoleBinding.Namespace}, foundSrb)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("Creating Istio ServiceRoleBinding", "namespace", istioServiceRoleBinding.Namespace,
+				"name", istioServiceRoleBinding.Name)
+			err = r.Create(context.TODO(), istioServiceRoleBinding)
+			if err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	} else {
+		if !reflect.DeepEqual(istioServiceRoleBinding.Spec, foundSrb.Spec) {
+			foundSrb.Spec = istioServiceRoleBinding.Spec
+			log.Info("Updating Istio ServiceRoleBinding", "namespace", istioServiceRoleBinding.Namespace,
+				"name", istioServiceRoleBinding.Name)
+			err = r.Update(context.TODO(), foundSrb)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	istioGateway := &istionetworking.Gateway{
```

Comment:
Why do you need a gateway per namespace?

Reply:
Because gateway is namespace scoped.
IIUC each gateway only responsible for one service mesh, in our case it's roughly services in one namespace.

Reply:
https://istio.io/docs/reference/config/networking/v1alpha3/virtual-service/#VirtualService

You can define virtualservices bind to gateway in another namespace.

So I thought we only need one istio-ingressgatway service + one gateway resource per cluster?

Reply:
Seems not working on our current Istio version.
Maybe we can resolve it in future by updating Istio to latest release.

Reply:
@kunmingg can you please share the manifest you are trying?
I found istio/istio#11131 and it seems you also need to specify the gateway's namespace. 
The syntax in new releases is <namespace>/<name>, but in older releases it was shortname or fqdn (<name>.<namespace>)
We should be able to make it work!

Reply:
@yanniszark 
I tried both:
- `<namespace>/<name>`: got `istio "gateway" configuration is invalid: domain name invalid`
- `<name>.<namespace>`: cannot access notebook from UI after change gateway.

The Istio I use is in https://github.com/kubeflow/kubeflow/tree/master/dependencies/istio

Reply:
@kunmingg I read the source code from the version we use and it seems that `/` notation is not introduced yet.
After trying a couple of different ways, I managed to make it work with `gateway: <name>.<namespace>.svc.cluster.local`.

However, we are using a pre-release version of Istio, which is by-definition not stable.
It seems to me that we should upgrade to a stable version.

Reply:
Great! Let me have a try.
Will update Istio to stable release.

Reply:
I updated the istio manifest to 1.1.6 in https://github.com/kubeflow/kubeflow/pull/3262

Reply:
Since we updated, the `<ns>/<name>` form will work and is the recommended way.